docs/internal/bimonad.md

```diff
@@ -0,0 +1,353 @@
+
+Rest of the structure defined as follows
+```text
+id{(a, x)} = (id{a}, id{x})
+(f, u) ∘ (g, v) = (f ∘ g, u ∘ v)
+```
+And properties are easy to prove.
+
+### Bifunctor
+
+Bifunctor is just a word meaning "functor from the product of categories", so bifunctor `C × D -> E` is just a functor `C × D -> E` 
+
+Technically that means that we have a binary function on objects and 
+mapping function, that mapping couple of functions into a single one.
+
+So bifunctor `Scala × Scala -> Scala` would have following interface 
+```scala
+trait Bifunctor[F[_, _]]{
+    def bimap[A, B, C, D](f: A => B, g: C => D): F[A, B] => F[C, D]
+}
+```
+
+One notable property of bifunctors is orthogonality of mappings, meaning
+
+```
+(f, id) ∘ (id, g) = (f ∘ id, id ∘ g) = (f, g) = (id ∘ f, g ∘ id) = 
+    (id, g) ∘ (f, id)
+```
+and for bifunctor `F` it would mean that 
+
+```text
+map{F}(f, id) ∘ map{F}(id, g) = map{F}(id, g) ∘ map{F}(f, id)
+```
+Which looks like some form of commutativity, but is not since `id`s have different implicit indices
+
+But it allows to define bifunctor in the following form
+```scala 
+trait Bifunctor[F[_, _]]{
+    def lmap[A, B, C](f: A => B): F[A, C] => F[B, C]
+    def rmap[A, B, C](f: A => B): F[C, A] => F[C, A]
+}
+```
+So we definitely may split any mappings to left and right components
+
+### Bifunctor monad
+
+So what is a monad which is also bifunctor, it should be endofunctor from a product of some categories, that means that it's a functor which results also in a product of the same categories,
+
+```
+pure: ∀a: C, x: D, Hom((a, x), F(a, x))
+flatMap: ∀(a, b: C), (x, y: D), Hom((a, x), F(b, y)) -> Hom(F(a, x), F(b, y))
+```
+
+But here we must remember that F is a function to a pair of objects.  
+How can we represent such thing on scala types?  
+We can immediately think of any function, which results in a pair as a pair of functions, producing corresponding elements of pair.  
+Meaning that `F: A -> B × C` can be represented `F_1: A -> B` and `F_2: A -> C` and `F(a) = (F_1(a), F_2(a))`.  
+So a binary function from types to a pair of types we can represent as a pair of type constructors.
+Same trick will work for components of `pure` and `flatMap` : instead of tuple of functions as a result we define two different methods for each component
+There for a bifunctor monad 
+`Scala × Scala -> Scala × Scala` would look as 
+```scala
+trait BiMonad[L[_, _], R[_, _]]{
+    def pureLeft[A, B](a: A): L[A, B]
+    def pureRight[A, B](b: B): R[A, B]
+    def flatMapLeft[A, B, C, D](fl: A => L[C, D], fr: B => R[C, D]): L[A, B] => L[C, D]
+    def flatMapRight[A, B, C, D](fl: A => L[C, D], fr: B => R[C, D]): R[A, B] => R[C, D]
+}
+```
+We can easily uncurry that and define bifunctors for both `L` and `R` using such definitions.
+
+### Twin monad
+
+Such definitions are still unsettling. It is hard to imagine how it could be useful to have two different type-constructors. So first thing we can narrow our definition requiring that 
```

Comment:
```suggestion
Such definitions are still unsettling. It is hard to imagine how it could be useful having two different type-constructors. So the most obvious way of narrowing our definition is requiring that 
```